osu.Game.Rulesets.Taiko/Mods/TaikoModSingleTap.cs

```diff
@@ -0,0 +1,136 @@
+        public override LocalisableString Description => @"One key for dons, one key for kats.";
+
+        protected bool CheckValidNewAction(TaikoAction action)
+        {
+            if (action == TaikoAction.LeftCentre || action == TaikoAction.RightCentre)
+                return LastAcceptedDonAction == null || LastAcceptedDonAction == action;
+            if (action == TaikoAction.LeftRim || action == TaikoAction.RightRim)
+                return LastAcceptedKatAction == null || LastAcceptedKatAction == action;
+
+            return true;
+        }
+
+        public override double ScoreMultiplier => 1.0;
+        public override Type[] IncompatibleMods => new[] { typeof(ModAutoplay), typeof(ModRelax), typeof(TaikoModCinema) };
+        public override ModType Type => ModType.Conversion;
+
+        private DrawableTaikoRuleset ruleset = null!;
+
+        private TaikoPlayfield playfield { get; set; } = null!;
+
+        protected TaikoAction? LastAcceptedDonAction { get; private set; }
+        protected TaikoAction? LastAcceptedKatAction { get; private set; }
+
+        /// <summary>
+        /// A tracker for periods where alternate should not be forced (i.e. non-gameplay periods).
+        /// </summary>
+        /// <remarks>
+        /// This is different from <see cref="Player.IsBreakTime"/> in that the periods here end strictly at the first object after the break, rather than the break's end time.
+        /// </remarks>
+        private PeriodTracker nonGameplayPeriods = null!;
+
+        private IFrameStableClock gameplayClock = null!;
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<TaikoHitObject> drawableRuleset)
+        {
+            ruleset = (DrawableTaikoRuleset)drawableRuleset;
+            ruleset.InputManager.Add(new InputInterceptor(this));
+            playfield = (TaikoPlayfield)ruleset.Playfield;
+
+            var periods = new List<Period>();
+
+            if (drawableRuleset.Objects.Any())
+            {
+                periods.Add(new Period(int.MinValue, getValidJudgementTime(ruleset.Objects.First()) - 1));
+
+                foreach (BreakPeriod b in drawableRuleset.Beatmap.Breaks)
+                    periods.Add(new Period(b.StartTime, getValidJudgementTime(ruleset.Objects.First(h => h.StartTime >= b.EndTime)) - 1));
+
+                static double getValidJudgementTime(HitObject hitObject) => hitObject.StartTime - hitObject.HitWindows.WindowFor(HitResult.Meh);
+            }
+
+            nonGameplayPeriods = new PeriodTracker(periods);
+
+            gameplayClock = drawableRuleset.FrameStableClock;
+        }
+
+        public void Update(Playfield playfield)
+        {
+            if (!nonGameplayPeriods.IsInAny(gameplayClock.CurrentTime)) return;
+
+            if (LastAcceptedDonAction != null)
+                LastAcceptedDonAction = null;
+
+            if (LastAcceptedKatAction != null)
+                LastAcceptedKatAction = null;
```

Comment:
Nit: Null checks serve no purpose and can be removed (as in, unconditionally assign `null`)